k4FWCore/components/PodioInput.cpp

```diff
@@ -184,16 +184,29 @@ PodioInput::PodioInput(const std::string& name, ISvcLocator* svcLoc) : Consumer(
     error() << "Could not get PodioDataSvc" << endmsg;
   }
   fillReaders();
-
-  auto key = edm4hep::EventHeaderName;
-  if (std::find(m_collectionNames.begin(), m_collectionNames.end(), key) == m_collectionNames.end()) {
-    m_collectionNames.value().push_back(key);
-  }
 }
 
 void PodioInput::operator()() const {
-  for (auto& collName : m_collectionNames) {
+  if (m_podioDataSvc->hasCollection(edm4hep::EventHeaderName)) {
+    m_readers[edm4hep::EventHeaderCollection::typeName](edm4hep::EventHeaderName);
+  } else {
+    info() << "No EventHeader collection found in the event. Not reading it" << endmsg;
+  }
+
+  const auto& collsToRead = [&]() {
+    if (m_collectionNames.empty()) {
+      return m_podioDataSvc->getEventFrame().getAvailableCollections();
+    } else {
+      return m_collectionNames.value();
+    }
+  }();
```

Comment:
my c++ knowledge is not as advanced but why do we need this as a lambda? :woozy_face:

Reply:
The pattern is called an *immediately invoked function expression (IIFE)* and it is mainly used for "complicated" initialization logic, where it can help to
- reduce the scope of temporary variables that are necessary for the initialization
- **Allows to declare the initialized object as `const`**, instead of having it mutable and dealing with the logic outside. 

 So in this case really only to have a `const&` instead of just a `&`.